app/recipe/views.py

```diff
@@ -1,26 +1,68 @@
 """
 Views for recipe API
+
+quite useful: https://testdriven.io/blog/drf-views-part-3/
+
 """
 from rest_framework import status
 from rest_framework.response import Response
-from rest_framework.viewsets import ModelViewSet
+from rest_framework.viewsets import ViewSet
+from django.shortcuts import get_object_or_404
 from .serializers import RecipeSerializer
-from .models import Recipe
-
-class RecipeViewSet(ModelViewSet):
-    queryset = Recipe.objects.all()
-    serializer_class = RecipeSerializer
-
-    # Overriding list...
-    # Search view by 'name' substring
-    def list(self, request, *args, **kwargs):
-        name = self.request.query_params.get('name', None)
-        queryset = self.queryset
-        if name:
-            queryset = queryset.filter(name__icontains=name)
-            print(queryset)
-            serializer = RecipeSerializer(queryset, many=True)
-            return Response(serializer.data)
-        else:
-            # Retrieve all recipes
-            return super().list(request, *args, **kwargs)
+from .services import RecipeService
+from .api import RecipeApi
+
+
+class RecipeViewSet(ViewSet):
+
+    def list(self, request):
+        try:
+            # recipes_dtos = RecipeService.get_all_recipes()
+            recipes_dtos = RecipeApi.get_all_recipes()
+            recipes_dict = [recipe_dto.dict() for recipe_dto in recipes_dtos]
+            return Response(data=recipes_dict, status=status.HTTP_200_OK)
+        except:
+            return Response("No recipes to show", status=status.HTTP_404_NOT_FOUND)
+
+    def retrieve(self, request, pk=None):
+        try:
+            # recipe_dto = RecipeService.get_recipe(pk) # e.g. aliasviewsets.py line 20
+            recipe_dto = RecipeApi.get_recipe(pk)
+            return Response(data=recipe_dto.dict(), status=status.HTTP_200_OK)
+        except:
+            return Response("Recipe not found.", status=status.HTTP_404_NOT_FOUND)
+
+
+    def create(self, request):
+        serializer = RecipeSerializer(data=request.data)
+        if serializer.is_valid():
+            # recipe_dto = RecipeService.create_recipe(serializer.validated_data)
+            # return Response(data=recipe_dto.dict(), status=status.HTTP_201_CREATED)
+            RecipeApi.create_recipe(serializer.validated_data)
+            return Response("Recipe successfully created.", status=status.HTTP_201_CREATED)
+        return Response("Recipe not created.", status=status.HTTP_400_BAD_REQUEST)
```

Comment:
We often wouldn't use the serializer like this but it's also fine. You're effectively using the serializer to make sure the data is valid for the model and then creating it if it is so all good